django/core/exceptions.py

```diff
@@ -182,6 +184,30 @@ def __str__(self):
     def __repr__(self):
         return 'ValidationError(%s)' % self
 
+    def __eq__(self, other):
+        if not isinstance(other, ValidationError):
+            return NotImplemented
+        if hasattr(self, 'message'):
+            return (
+                hasattr(other, 'message') and
+                self.message == other.message and
+                self.code == other.code and
+                self.params == other.params
+            )   
+        return (
+            hasattr(self, 'error_list') == hasattr(other, 'error_list') and
+            hasattr(self, 'error_dict') == hasattr(other, 'error_dict') and
+            sorted(self.messages) == sorted(other.messages)
+        )
+
+    def __hash__(self):
+        if hasattr(self, 'message'):
+            return hash((make_hashable(self.message), make_hashable(self.code), make_hashable(self.params)))
```

Comment:
You don't need to use `make_hashable()` for primitives:
```suggestion
            return hash((self.message, self.code, self.params))
```

Reply:
`self.params` is a dict? (I'll change it for message + code)

``` python
    return hash((self.message, self.code, self.params))
TypeError: unhashable type: 'dict'
```

Reply:
so maybe `hash((self.message, self.code, make_hashable(self.params)))` :thinking:

Reply:
I agree 👍